common/src/main/kotlin/streams/service/errors/KafkaDLQService.kt

```diff
@@ -0,0 +1,85 @@
+package streams.service.errors
+
+import org.apache.commons.lang3.exception.ExceptionUtils
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.Producer
+import org.apache.kafka.clients.producer.ProducerConfig
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.record.RecordBatch
+import org.apache.kafka.common.serialization.ByteArraySerializer
+import org.neo4j.util.VisibleForTesting
+import java.util.*
+
+class KafkaDLQService(private val producer: Producer<ByteArray, ByteArray>?, private val errorConfig: ErrorConfig, private val log: (String,Exception?)->Unit): ErrorService() {
+
+    constructor(config: Properties, errorConfig: ErrorConfig,
+                log: (String, Exception?) -> Unit) : this(producer(errorConfig, config), errorConfig, log)
+
+    companion object {
+        private fun producer(errorConfig: ErrorConfig, config: Properties) =
+                errorConfig.dlqTopic?.let {
+                    val props = Properties()
+                    props.putAll(config)
+                    props[ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG] = ByteArraySerializer::class.java.name
+                    props[ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG] = ByteArraySerializer::class.java.name
+                    KafkaProducer<ByteArray,ByteArray>(props)
+                }
+    }
+
+    override fun report(errorDatas: List<ErrorData>) {
+        if (errorConfig.fail) throw ProcessingError(errorDatas)
```

Comment:
The core conditional for what kind of error reporting to do is inside of the DLQ service.  But there might not be a DLQ.   So do I get this right the design we're going for is that you'll always have a DLQ service, just that if the topic isn't configured, the printing is done by the DLQ service?

This is OK but design wise seems somewhat strange; it seems like this might be better an "ErrorService" rather than a DLQ service.  Just one of its behaviors involves a DLQ.

I think the shape of this object is OK but it might not be named properly anymore.

Reply:
Yes, true. I'll change that.